static/app/utils/profiling/renderers/textRenderer.tsx

```diff
@@ -0,0 +1,164 @@
+import {
+  ELLIPSIS,
+  findRangeBinarySearch,
+  getContext,
+  Rect,
+  resizeCanvasToDisplaySize,
+  trimTextCenter,
+} from '../gl/utils';
+
+export function isOutsideView(frame: Rect, view: Rect, inverted: boolean): boolean {
+  // Frame is outside of the view on the left
+  if (frame.right <= view.left) {
+    return true;
+  }
+
+  // Frame is outside of the view on the right
+  if (frame.left >= view.right) {
+    return true;
+  }
+
+  // Frame is above the view
+  if (frame.bottom <= view.top) {
+    return true;
+  }
+
+  // Frame is below the view
+  if (frame.top >= view.bottom) {
+    return true;
+  }
+
+  // @TODO check if we still need this
+  if (inverted) {
+    if (frame.top - 1 >= view.bottom) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+class TextRenderer {
+  textCache: Record<string, number> = {};
+
+  canvas: HTMLCanvasElement;
+  context: CanvasRenderingContext2D;
+  theme: FlamegraphTheme;
+  flamegraph: Flamegraph;
+
+  constructor(canvas: HTMLCanvasElement, flamegraph: Flamegraph, theme: FlamegraphTheme) {
+    this.canvas = canvas;
+    this.theme = theme;
+    this.flamegraph = flamegraph;
+
+    this.context = getContext(canvas, '2d');
+
+    resizeCanvasToDisplaySize(canvas);
+  }
+
+  clearCache(): void {
+    this.textCache = {};
+  }
+
+  measureText(context: CanvasRenderingContext2D, text: string): number {
+    if (this.textCache[text]) {
+      return this.textCache[text];
+    }
+    this.textCache[text] = context.measureText(text).width;
+    return this.textCache[text];
+  }
+
+  draw(configViewSpace: Rect, configSpace: Rect, configToPhysicalSpace: mat3): void {
+    this.context.font = `${
+      this.theme.SIZES.BAR_FONT_SIZE * window.devicePixelRatio
+    }px "Source Code Pro", Courier, monospace`;
+
+    this.context.textBaseline = 'alphabetic';
+    this.context.fillStyle = this.theme.COLORS.LABEL_FONT_COLOR;
+
+    const minWidth = this.measureText(this.context, `${ELLIPSIS}`);
+
+    let frame: FlamegraphFrame;
+    let i: number;
+    const length: number = this.flamegraph.frames.length;
+
+    // We currently iterate over all frames, but we could optimize this to only iterate over visible frames.
+    // This could be achieved by querying the flamegraph tree (as an interval tree). This would still run in O(n), but
+    // would improve our best case performance (e.g. when users) zoom into the flamegraph
+    for (i = 0; i < length; i++) {
+      frame = this.flamegraph.frames[i];
+
+      // This rect gets discarded after each render which is wasteful
+      const frameInConfigSpace = new Rect(
+        frame.start,
+        this.flamegraph.inverted ? configSpace.height - frame.depth + 1 : frame.depth,
+        frame.end - frame.start,
+        1
+      );
+
+      // Check if our rect overlaps with the current viewport and skip it
+      if (
+        isOutsideView(frameInConfigSpace, configViewSpace, !!this.flamegraph.inverted)
```

Comment:
Maybe this is because I'm not sure what `inverted` means, but if `inverted` is already used to initialize the `Rect`, it shouldn't be needed for this `isOutsideView` check right?

Reply:
Yes, we dont initialize inverted with Rect right now because I dont want to tie it to the flamegraph. The problem is that currently we kind of fake invert rectangles... we remap their depth, whereas the real solution would be to change coordinate system. That could be a good task if you want to take a stab at it, it would allow us to simplify some of the code and make it more straightforward